### PR TITLE
More than max_no_competitors could be created in network games

### PR DESCRIPTION
**Pull request by SamuXarick**
Description: There is a max number of commands_per_frame that can cause the starting of a company to be delayed by another tick. I was seeing 5 AIs being started when I've set max_no_competitors only to 4.

https://github.com/OpenTTD/OpenTTD/commit/011257dc8804175dd7d1e839e97e796c0a88aee6 causes it.

### CODE REVIEW

File: src/company_cmd.cpp [hunk 1]
```diff
@@ -884,6 +885,22 @@ CommandCost CmdCompanyCtrl(TileIndex tile, DoCommandFlag flags, uint32 p1, uint3
...
+			if (GB(p2, 0, 1) == 1) {
+				Company *c;
+
+				/* Count number of competitors */
+				uint n = 0;
+				FOR_ALL_COMPANIES(c) {
+					if (c->is_ai) n++;
+				}
+
+				if (n >= (uint)_settings_game.difficulty.max_no_competitors) return CommandCost(); // can't return CMD_ERROR in this case
```
Comment (glx22) [inline src/company_cmd.cpp]: Why can't you return CMD_ERROR ?

**Comment by SamuXarick** [comment]
Description: There is a max number of commands per frame that can cause the starting of a company to be delayed by another tick. I was seeing 5 AIs being started when I've set max_no_competitors only to 4.

File: src/company_cmd.cpp [hunk 2]
```diff
@@ -818,6 +818,7 @@ void CompanyAdminRemove(CompanyID company_id, CompanyRemoveReason reason)
  * - bits 16..24: CompanyID
  * @param p2 various depending on CompanyCtrlAction
  * - bits 0..31: ClientID (with CCA_NEW)
+ * - bit 0: If set to 1, enforce the max_no_competitors test (with CCA_NEW_AI)
```
Comment (PeterN) [inline src/company_cmd.cpp]: Why would you ever want to not enforce this?

File: src/company_cmd.cpp [hunk 3]
```diff
@@ -881,6 +882,21 @@ CommandCost CmdCompanyCtrl(TileIndex tile, DoCommandFlag flags, uint32 p1, uint3
 		}
 
 		case CCA_NEW_AI: { // Make a new AI company
+#ifdef ENABLE_NETWORK
```
Comment (PeterN) [inline src/company_cmd.cpp]: Please update to remove this, not sure why you'd want the behaviour to differ based on how the game is **compiled** anyway.

File: src/company_cmd.cpp [hunk 4]
```diff
@@ -881,6 +882,21 @@ CommandCost CmdCompanyCtrl(TileIndex tile, DoCommandFlag flags, uint32 p1, uint3
...
+			/* Enforce the max_no_competitors test */
+			bool enforce = HasBit(p2, 0);
+			if (enforce) {
```
Comment (PeterN) [inline src/company_cmd.cpp]: Why should this test be optional?

File: src/company_cmd.cpp [hunk 5]
```diff
@@ -612,7 +612,7 @@ static bool MaybeStartNewCompany()
 	if (n < (uint)_settings_game.difficulty.max_no_competitors) {
 		/* Send a command to all clients to start up a new AI.
 		 * Works fine for Multiplayer and Singleplayer */
-		return DoCommandP(0, CCA_NEW_AI | INVALID_COMPANY << 16, 0, CMD_COMPANY_CTRL);
+		return DoCommandP(0, CCA_NEW_AI | INVALID_COMPANY << 16, 1, CMD_COMPANY_CTRL);
```
Comment (PeterN) [inline src/company_cmd.cpp]: Possibly unnecessary given my other comments

**Review by PeterN — CHANGES_REQUESTED** [review]

File: src/company_cmd.cpp (hunk 2 above)
Comment (SamuXarick) [reply src/company_cmd.cpp]: There are currently some places in the code where you start an ai company disregarding the max_no_competitors, like reload ai, startai console command

Comment (PeterN) [reply src/company_cmd.cpp]: Okay but why? If it bypasses max_no_competitors, what's the point of max_no_competitors?

Comment (SamuXarick) [reply src/company_cmd.cpp]: The point is to limit the number of AIs that can start automatically. Manually starting AIs past the limit is okay in my view, You are in control of how you want that to happen and when. It has always been like that.

**Comment by TrueBrain** [comment]
This feels to me like the wrong approach. From what I understand.
- The server creates a new AI; this causes a command to be queued
- Clients + server receive this command; they validate again if the server should have said this. If the server clearly should not have said this, everyone ignores the command.

This feels like a very dirty hack. The clients should never ignore a command from the server like that. It feels this needs more thinking, and possibly a better approach to resolve.

**Review by glx22 — CHANGES_REQUESTED** [review]
Wrong commit reference, should be 011257d. Please edit the commit message.

**Review by glx22 — DISMISSED** [review]

**Comment by SamuXarick** [comment]
> This feels to me like the wrong approach. From what I understand.
> 
> The server creates a new AI; this causes a command to be queued
> Clients + server receive this command; they validate again if the server should have said this. If the server clearly should not have said this, everyone ignores the command.
> 
> This feels like a very dirty hack. The clients should never ignore a command from the server like that. It feels this needs more thinking, and possibly a better approach to resolve.

This is a problem that happens in multiplayer. Single player does not apply.

Problem is difficult to follow, but gonna try explain with an example.

In multiplayer, when multiple AIs are set to start with 0 days delay between each other, OnTick_Companies() is called every tick to start them. Let's imagine I've set max_no_competitors to 5 and network.max_companies to 15.

Tick 1: 
On the first tick, OnTick_Companies() is called, MaybeStartNewCompany() is tested and it will give Company::GetNumItems() == 0 (assuming dedicated server), and will count the number of AIs, n == 0. Both tests fail, which allows DoCommandP to be sent.

This command will run in DC_TEST mode. CmdCompanyCtrl returns with CommandCost(). Since this is a network, the command is sent into a queue, it's not executed right away. It does a NetworkSendCommand and puts it in _local_wait_queue, then returns Succeeded.

For the rest of this tick, nothing happens. Game proceeds to next tick.

Tick 2:
NetworkDistributeCommands() is called, which runs DistributeQueue(). There is a commands_per_frame setting there, set to 2. I see it moving the command from _local_wait_queue into _local_execution_queue via DistributeCommandPacket. Since there's only 1 command in _local_wait_queue for now, commands_per_frame limit is not reached.

After that splitting, NetworkExecuteLocalCommandQueue() is called. The command from _local_execution_queue is run. CmdCompanyCtrl is first run in DC_TEST mode, returning with CommandCost(), and then it is run again in DC_EXEC mode, where it actually creates the AI company.

Now the cycle repeats, OnTick_Companies() is called, MaybeStartNewCompany() is tested and it will give Company::GetNumItems() == 1, and will count the number of AIs, n == 1. Both tests fail, which allows DoCommandP to be sent.

This command will run in DC_TEST mode. CmdCompanyCtrl returns with CommandCost(). Since this is a network, the command is sent into a queue, it's not executed right away. It does a NetworkSendCommand and puts it in _local_wait_queue, then returns Succeeded.

For the rest of this tick, the first AI that started does its stuff and also queues one command in _local_wait_queue. So _local_wait_queue now has 2 commands.

Tick 3:
NetworkDistributeCommands() is called, which runs DistributeQueue(). It is now moving 2 commands from _local_wait_queue to _local_execution_queue. commands_per_frame limit is reached, but since there was only 2 commands, _local_wait_queue still becomes empty.

NetworkExecuteLocalCommandQueue() will run the commands in _local_execution_queue in the order they were queued. First, CmdCompanyCtrl in DC_TEST mode and DC_EXEC mode, where the 2nd AI company is created. And second, the command that the 1st AI had queued.

The cycle repeats, OnTick_Companies() is called, MaybeStartNewCompany() is tested and it will give Company::GetNumItems() == 2, and will count the number of AIs, n == 2. Both tests fail, which allows DoCommandP to be sent.

CmdCompanyCtrl() is run in DC_TEST mode, then NetworkSendCommand puts it in _local_wait_queue and the command returns Succeeded.

For the rest of this tick, the first AI queues a command and the second AI also queues a command. _local_wait_queue now has 3 commands.

Tick 4:
NetworkDistributeCommands() is called, which runs DistributeQueue(). It is moving 2 commands from _local_wait_queue to _local_execution_queue. commands_per_frame limit is reached, but there were 3 commands, _local_wait_queue remains with one command, the last that was queued, which is the command from the 2nd AI.

NetworkExecuteLocalCommandQueue() will run the commands in _local_execution_queue in the order they were queued. First, CmdCompanyCtrl in DC_TEST mode and DC_EXEC mode, where the 3nd AI company is created. And second, the command that the 1st AI had queued. The command from the 2nd AI is not executed.

Repeat, OnTick_Companies() is called, MaybeStartNewCompany() is tested and it will give Company::GetNumItems() == 3, and will count the number of AIs, n == 3. Both tests fail, which allows DoCommandP to be sent.

CmdCompanyCtrl() is run in DC_TEST mode, then NetworkSendCommand puts it in _local_wait_queue and the command returns Succeeded. _local_wait_queue now contains 2 commands, the command from the 2nd AI, and creation of the 4th AI company.

For the rest of this tick, the first AI queues a command, the second AI does nothing, and the third AI queues a command. _local_wait_queue now has 4 commands: command from the 2nd AI, creation of the 4th AI company, command from 1st AI, and command from 3rd AI.

Tick 5:
NetworkDistributeCommands() moves 2 commands from _local_wait_queue to _local_execution_queue.

NetworkExecuteLocalCommandQueue() will run 2 commands from _local_execution_queue, first, the command from the 2nd AI, then CmdCompanyCtrl in DC_TEST, then DC_EXEC mode, creating the 4th AI company.

OnTick_Companies(), MaybeStartNewCompany()… Company::GetNumItems() == 4, n == 4. Both tests fail, which allows DoCommandP to be sent.

CmdCompanyCtrl() is run in DC_TEST mode, then NetworkSendCommand puts it in _local_wait_queue and the command returns Succeeded. _local_wait_queue now contains 3 commands, the command from the 1st AI and 3rd AIs, and creation of the 5th AI company.

For the rest of this tick, 2nd AI queues a command, and 4th AI also queues a command. _local_wait_queue contains 5 commands: command from 1st AI, command from 3rd AI, creation of 5th company, command from 2nd AI, and command from 4th AI.

Tick 6 (is where the problem first arises):
NetworkDistributeCommands() moves 2 commands from _local_wait_queue to _local_execution_queue. _local_wait_queue will remain with 3 commands, which are creation of 5th company, command from 2nd AI, and command from 4th AI.

NetworkExecuteLocalCommandQueue() will run 2 commands from _local_execution_queue, which are command from 1st AI and command from 3rd AI. This means the 5th company is not created this tick!

OnTick_Companies() happens, MaybeStartNewCompany() is doing tests and Company::GetNumItems() == 4, n == 4. It does not know that there was a 5th company! The tests in this situation are incorrect, but they will still fail, allowing the queue of the 6th AI company.

CmdCompanyCtrl() is run in DC_TEST mode, then NetworkSendCommand puts it in _local_wait_queue and the command returns Succeeded. _local_wait_queue now contains 4 commands: the creation of the 5th AI company, command from the 2nd AI, command from the 4th AI and creation of the 6th company.

For the rest of the tick, commands from 1st and 3rd AIs are queued. _local_wait_queue now contains 6 commands: the creation of the 5th AI company, command from the 2nd AI, command from the 4th AI, creation of the 6th company, command from 1st AI and command from 3rd AI.

Tick 7:
NetworkDistributeCommands() moves 2 commands from _local_wait_queue to _local_execution_queue. _local_wait_queue will remain with 4 commands which are command from the 4th AI, creation of 6th company, command from 1st AI and command from 3rd AI.

NetworkExecuteLocalCommandQueue() runs 2 commands from _local_execution_queue. First, CmdCompanyCtrl in DC_TEST and DC_EXEC modes. The 5th company is created. And then, the command from the 2nd AI is executed.

OnTick_Companies(), MaybeStartNewCompany() this time gets Company::GetNumItems() == 5 and n == 5. Since 5 was the limit for the max_no_competitors, the test will pass, allowing this function to return false. No more companies are created… but this comes too late. The 6th AI company is still in the queue, remember.

For the rest of the tick, commands from 2nd and 5th AI are queued. _local_wait_queue contains 6 commands: command from 4th AI, creation of 6th AI company, command from 1st AI, command from 3rd AI, command from 2nd AI and command from 5th AI.

Tick 8 (the final rest of the problem):
NetworkDistributeCommands() moves 2 commands from _local_wait_queue to _local_execution_queue. _local_wait_queue will remain with 4 commands: command from 1st AI, command from 3rd AI, command from 2nd AI and command from 5th AI.

NetworkExecuteLocalCommandQueue() runs 2 commands from _local_execution_queue. First, the command from the 4th AI is executed. And finally, CmdCompanyCtrl in DC_TEST and DC_EXEC modes. Without this patch, the 6th company is created. With it, both DC_TEST and DC_EXEC fail. No 6th AI company is created.

The rest doesn't matter anymore.

This big example may not truly reflect what happens in the code, it's just my mere observations following the code in visual studio. I have not really checked how different the execution is on a client other than the server. I assumed that, since _settings_client.network.max_companies is not a game setting, I decided to pass it through p2. Client and server could probably have different values for this setting, that's why I avoid testing against it.

**Comment by TrueBrain** [comment]
Tnx for the lengthy explanation. So basically what I said holds. So that means your solution is more a hack than a real solution, as you didn't tackle the real issue here.
Why I say it is a hack, if you look a bit further: say there are 100 commands in the queue, and you want 10 AIs. This would mean that for 20 ticks (100 / 5), 10 AI commands are queued. So by the time the first AI wants to be create, 20 * 10 commands more are in the queue. Of course 19 * 10 commands are later dismissed. This is what I mean with: this feels like a hack. You are not solving the issue, but an underlying symptom.

So we should look for how to solve this properly.

I think it all originates in the original patch: looping over a function to create multiple sounds like a dirty approach on its own. If you want to start 10 AIs in a single tick, there should be a single command creating those 10 AIs. Not loop over creating 1 AI 10 times, hoping it can be done in a single tick.
So possibly, instead of looking in this direction, you can look in that (or any other) direction?

**Comment by nielsmh** [comment]
Idea for making a single command to start multiple AIs: Use 16 bits of either p1 or p2 to CCA_NEW_AI as a bitmask of companies to start AI for. Then loop over all set bits in that mask inside the command execution.

File: src/ai/ai_core.cpp [hunk 6]
```diff
@@ -308,11 +308,13 @@
 	}
 }
 
-/* static */ int AI::GetStartNextTime()
+/* static */ int AI::GetStartNextTime(int ids_to_skip)
 {
-	/* Find the first company which doesn't exist yet */
+	/* Find the first company which doesn't exist yet, but skip some indexes */
```
Comment (TrueBrain) [inline src/ai/ai_core.cpp]: `some`. Just any amount? Randomly? This really could use some rewording. Especially I am missing WHY it is happening. The rest I can read in the code just fine.

File: src/company_cmd.cpp [hunk 7]
```diff
@@ -592,26 +592,29 @@ void StartupCompanies()
 	_next_competitor_start = 0;
 }
 
-/** Start a new competitor company if possible. */
-static bool MaybeStartNewCompany()
+/** Start new competitor companies if possible. */
+void MaybeStartNewCompany()
 {
-	if (_networking && Company::GetNumItems() >= _settings_client.network.max_companies) return false;
+	/* Allow multiple AIs to possibly start in the same tick. */
```
Comment (TrueBrain) [inline src/company_cmd.cpp]: Again this undefined wording ... `possibly` .. well, can you or can you not?

File: src/company_cmd.cpp [hunk 8]
```diff
@@ -592,26 +592,29 @@ void StartupCompanies()
...
-	FOR_ALL_COMPANIES(c) {
-		if (c->is_ai) n++;
-	}
+		/* count number of competitors */
+		uint n = num_to_start;
+		FOR_ALL_COMPANIES(c) {
+			if (c->is_ai) n++;
+		}
+		if (n >= (uint)_settings_game.difficulty.max_no_competitors) break;
+	} while (_settings_game.difficulty.max_no_competitors - ++num_to_start > 0 && AI::GetStartNextTime(num_to_start) == 0);
```
Comment (TrueBrain) [inline src/company_cmd.cpp]: This is really bad coding. Comparing, incrementing, and substracting in a single statement. Please don't punish future people like this ;)

File: src/company_cmd.cpp [hunk 9]
```diff
@@ -870,12 +867,16 @@ CommandCost CmdCompanyCtrl(TileIndex tile, DoCommandFlag flags, uint32 p1, uint3
...
 
-		case CCA_NEW_AI: { // Make a new AI company
+		case CCA_NEW_AI: { // Make new AI companies
 			if (!(flags & DC_EXEC)) return CommandCost();
 
 			if (company_id != INVALID_COMPANY && (company_id >= MAX_COMPANIES || Company::IsValidID(company_id))) return CMD_ERROR;
-			Company *c = DoStartupNewCompany(true, company_id);
-			if (c != NULL) NetworkServerNewCompany(c, NULL);
+
+			uint num_to_start = max(1U, GB(p2, 0, 4));
```
Comment (TrueBrain) [inline src/company_cmd.cpp]: I would say a value of 0 (as a value > `MAX_COMPANIES`) should return `CMD_ERROR`. Not silently be ignored.

File: src/company_cmd.cpp [hunk 10]
```diff
@@ -870,12 +867,16 @@ CommandCost CmdCompanyCtrl(TileIndex tile, DoCommandFlag flags, uint32 p1, uint3
...
+			do {
+				Company *c = DoStartupNewCompany(true, company_id);
+				if (c != NULL) NetworkServerNewCompany(c, NULL);
+			} while (--num_to_start > 0);
```
Comment (TrueBrain) [inline src/company_cmd.cpp]: `do/while` loops in general are very difficult to read. Why not a `while` loop here? Or better, a `for` loop?

File: src/openttd.cpp [hunk 11]
```diff
@@ -909,11 +909,14 @@ static void MakeNewGameDone()
 {
 	SettingsDisableElrail(_settings_game.vehicle.disable_elrails);
 
+	bool start_ai_companies = _game_mode != GM_MENU && AI::CanStartNew() && AI::GetStartNextTime() == 0;
```
Comment (TrueBrain) [inline src/openttd.cpp]: Is this really needed? Is a delay of 1 day really terible? All these exception codes are hard to maintain and difficult to track. Do we really want this?
If so, please do add a comment WHY this is here. As now it just is .. without much context.

**Review by TrueBrain — CHANGES_REQUESTED** [review]

**Comment by TrueBrain** [comment]
PS: please fix commit message length. And please please, never ever again make title of Pull Request this long please. It is not helping, and kinda makes me want to walk away ..

File: src/company_cmd.cpp (hunk 9 above)
Comment (SamuXarick) [reply src/company_cmd.cpp]: Uh, that means I have to change p2 everywhere to 1.

Comment (TrueBrain) [reply src/company_cmd.cpp]: Sounds you are being lazy :)

File: src/company_cmd.cpp [hunk 12]
```diff
@@ -592,26 +592,29 @@ void StartupCompanies()
...
-	/* count number of competitors */
-	uint n = 0;
-	FOR_ALL_COMPANIES(c) {
-		if (c->is_ai) n++;
-	}
+		/* count number of competitors */
+		uint n = num_to_start;
+		FOR_ALL_COMPANIES(c) {
+			if (c->is_ai) n++;
+		}
```
Comment (TrueBrain) [inline src/company_cmd.cpp]: didn't even see this: there is no need to iterate over all companies inside this loop; you can just do it outside, and use the outcome of the loop inside the loop. Most likely makes a bit cleaner code too.

File: src/company_cmd.cpp [hunk 13]
```diff
@@ -592,26 +592,32 @@ void StartupCompanies()
...
+		if (c->is_ai) num_ais++;
 	}
 
-	if (n < (uint)_settings_game.difficulty.max_no_competitors) {
-		/* Send a command to all clients to start up a new AI.
-		 * Works fine for Multiplayer and Singleplayer */
-		return DoCommandP(0, CCA_NEW_AI | INVALID_COMPANY << 16, 0, CMD_COMPANY_CTRL);
+	/* Count number of AI companies to start. */
+	uint num_to_start;
+	for (num_to_start = 0; _settings_game.difficulty.max_no_competitors - num_to_start > 0 && AI::GetStartNextTime(num_to_start) == 0; num_to_start++) {
```
Comment (TrueBrain) [inline src/company_cmd.cpp]: you might want to check on this statement now.
`_settings_game.difficulty.max_no_competitors - num_to_start > 0` vs `num_ais + num_to_start >= (uint)_settings_game.difficulty.max_no_competitors`. Sounds one of them is redundant.
`AI::GetStartNextTime(num_to_start) == 0` <- possibly better to put this in an `if()` statement too, with a comment to explain why you want to loop again if this is zero.

File: src/ai/ai.hpp [hunk 14]
```diff
@@ -137,8 +137,9 @@ class AI {
 
 	/**
 	 * Get the number of days before the next AI should start.
+	 * @param ids_to_skip Number of company indexes to skip, that are pretended to have been started, but actually aren't yet.
```
Comment (TrueBrain) [inline src/ai/ai.hpp]: This sentence is hard to parse. Especially so, because the function is `GetStartNextTime`, so .. it feels a bit weird. It needs more context of some sorts. Possible suggestion (no clue if it is any good):
"For which Nth AI to retrieve the start_date.".
Possibly that needs a rename of the variable too.

**Review by TrueBrain — CHANGES_REQUESTED** [review]

File: src/company_cmd.cpp (hunk 13 above)
Comment (SamuXarick) [reply src/company_cmd.cpp]: Sorry, I reverted back to do/while cycle again, as my for cycle was bugged. I don't know how to do it, need a bit of help.
https://paste.openttdcoop.org/pixwxtrxo

File: src/ai/ai.hpp [hunk 15]
```diff
@@ -137,8 +137,10 @@ class AI {
 
 	/**
 	 * Get the number of days before the next AI should start.
+	 * @param count Current count of number of AI company indexes that are planned to be starting.
```
Comment (TrueBrain) [inline src/ai/ai.hpp]: This comment still doesn't make any sense. See my last comment.

It only makes sense if you also know about how it is being used. Which is never how comments work. Say someone is going to use this function to tell when each AI is going to start. Than this comment is totally out of place.

Please don't ignore comments people give in reviews. At least reply on them if you don't agree, which is fine, but ignoring is never an option.

**Review by TrueBrain — CHANGES_REQUESTED** [review]

File: src/openttd.cpp (hunk 11 above)
Comment (TrueBrain) [reply src/openttd.cpp]: Still no answer.

**Comment by TrueBrain** [comment]
I added two commits to your branch as suggestion for what you can do to make it, in my opinion, a bit more readable and understandable.

**Comment by TrueBrain** [comment]
We are reverting the original change from 1.9, giving more room in 1.10 to fix this properly and in time.

File: src/ai/ai.hpp (hunk 15 above)
Comment (SamuXarick) [reply src/ai/ai.hpp]: I wasn't ignoring (or maybe I was). I just couldn't figure if your comment would suffice. Pretty vague statement, so that's why I didn't say anything.

File: src/openttd.cpp (hunk 11 above)
Comment (SamuXarick) [reply src/openttd.cpp]: This is outdated, it was wrong.

File: src/openttd.cpp [hunk 16]
```diff
@@ -947,6 +950,10 @@ static void MakeNewGameDone()
 
 	CheckEngines();
 	CheckIndustries();
+
+	/* Check if the first AI is set to start immediately */
```
Comment (SamuXarick) [inline src/openttd.cpp:921]: This is done after creating the human company which takes the first AI line in openttd.cfg. The AI that is started here is the second in openttd.cfg. I don't know if it's fine to say "the first AI".

Comment (TrueBrain) [reply src/openttd.cpp:921]: huh? What does openttd.cfg have to do with any of this? Sometimes you talk in riddles.

File: src/ai/ai.hpp [hunk 17]
```diff
@@ -137,8 +137,9 @@ class AI {
 
 	/**
 	 * Get the number of days before the next AI should start.
+	 * @param count For which Nth AI to retrieve the starting date.
```
Comment (SamuXarick) [inline src/ai/ai.hpp:138]: I still don't know what to do here. Feels like something is missing. GetStartNextTime with a parameter has a very specific use for that loop in MaybeStartNewCompany. It is called in the middle of it. For each ai that is to start, ais_to_start count increases. When checking for the next AI to start, it has to skip those that were already accounted for.

"For which Nth AI to retrieve the starting date." - I don't know, because when looking at the function, it is only looking for non valid companies where an AI could possibly start, and retrieves the start_date of that position. But the position is not necessarily ordered. "Nth" feels wrong.

File: src/company_cmd.cpp [hunk 18]
```diff
@@ -870,12 +873,18 @@ CommandCost CmdCompanyCtrl(TileIndex tile, DoCommandFlag flags, uint32 p1, uint3
...
 		}
 
-		case CCA_NEW_AI: { // Make a new AI company
+		case CCA_NEW_AI: { // Make new AI companies
+			uint num_to_start = GB(p2, 0, 4);
+			if (num_to_start == 0) return CMD_ERROR;
+
 			if (!(flags & DC_EXEC)) return CommandCost();
 
 			if (company_id != INVALID_COMPANY && (company_id >= MAX_COMPANIES || Company::IsValidID(company_id))) return CMD_ERROR;
```
Comment (glx22) [inline src/company_cmd.cpp]: Should probably be moved above the DC_EXEC check, and modified to enforce INVALID_COMPANY if num_to_start > 1

File: src/openttd.cpp (hunk 16 above)
Comment (SamuXarick) [reply src/openttd.cpp:921]:
```
[ai_players]
none = start_date=0
"LuDiAI AfterFix" = air_support=1,cities_only=0,is_friendly=1,pf_profile=0,pick_mode=3,road_cap_mode=2,road_min_dist=130,road_support=1,scp_support=0,select_town_cargo=0,start_date=0,station_spread=0
RailwAI = plant_trees=10,start_date=0,use_trams=0
CivilAI = MaxBus=500,MaxCar=0,MinAirRange=256,MinPop=1500,NetworkRadius=4096,ShipRange=128,TrainRange=256,start_date=0
"LuDiAI AfterFix" = is_friendly=1,pf_profile=0,pick_mode=0,road_cap_mode=1,start_date=0
SimpleAI = start_date=0
SimpleAI = start_date=0
SimpleAI = start_date=0
SimpleAI = start_date=0
SimpleAI = start_date=0
SimpleAI = start_date=0
SimpleAI = start_date=0
SimpleAI = start_date=0
SimpleAI = start_date=0
SimpleAI = start_date=0
```
Talking about this.
Maybe first available AI would be better.

Comment (SamuXarick) [reply src/openttd.cpp:921]: You know what? Nevermind.

File: src/ai/ai.hpp (hunk 17 above)
Comment (SamuXarick) [reply src/ai/ai.hpp:138]: Nevermind

**Comment by LordAro** [comment]
Given 011257d has long since been reverted, what is this PR doing, exactly?

**Comment by SamuXarick** [comment]
LordAro, #7376 is trying to generate the correct number of AIs in multiplayer. Just tested 20191230-master-gf6ce5c4563 and the bug is still present.
Unnamed, 1950-01-0116

File: src/company_cmd.cpp [hunk 19]
```diff
@@ -706,19 +722,13 @@ void OnTick_Companies()
 	}
 
 	if (_next_competitor_start == 0) {
-		/* AI::GetStartNextTime() can return 0. */
-		_next_competitor_start = max(1, AI::GetStartNextTime() * DAY_TICKS);
+		/* If GetStartNextTime is zero, wait at least a day to re-evaluate.
+		 * This avoids checking every tick if an AI can be started. */
+		_next_competitor_start = max(DAY_TICKS, AI::GetStartNextTime() * DAY_TICKS);
```
Comment (SamuXarick) [inline src/company_cmd.cpp:727]: Can I revert this line? The whole point of this PR was so that it could create the correct number of AIs while checking every tick.

**Comment by LordAro** [comment]
Going to close this one. It's not been touched in a while, and there's been no real consensus on it, other than "this is the wrong solution"
